chatkit-core/src/main/kotlin/com/pusher/chatkit/presence/PresenceService.kt

```diff
@@ -9,22 +9,40 @@ import java.net.URLEncoder
 import java.util.*
 
 class PresenceService(
-        myUserId: String,
+        private val myUserId: String,
         private val client: PlatformClient,
         private val consumer: PresenceSubscriptionConsumer,
         private val logger: Logger
 ) {
     private val subscriptions = HashMap<String, Subscription>()
+    private var registrationSub: Subscription? = null
 
-    private val registrationSub =
-            loggingSubscription(
-                    path = "/users/${URLEncoder.encode(myUserId, "UTF-8")}/register",
-                    listeners = SubscriptionListeners(),
-                    messageParser = PresenceSubscriptionEventParser(myUserId),
-                    logger = logger,
-                    client = client,
-                    description = "PresenceRegistration $myUserId"
-            )
+    init {
+        goOnline()
+    }
+
+    fun goOnline() {
+        synchronized(this) {
+            if (registrationSub == null) {
+                registrationSub =
+                        loggingSubscription(
+                                path = "/users/${URLEncoder.encode(myUserId, "UTF-8")}/register",
+                                listeners = SubscriptionListeners(),
+                                messageParser = PresenceSubscriptionEventParser(myUserId),
+                                logger = logger,
+                                client = client,
+                                description = "PresenceRegistration $myUserId"
+                        )
+            }
+        }
+    }
+
+    fun goOffline() {
```

Comment:
I wonder if this implies enough that you're no longer going to get any presence-related updates. Some might think that this would allow you to do an "invisible" mode, but that's not the case.

Reply:
It's not public yet. It it's made public, it would be called "invisible", as you clearly are online.

But it also probably won't be make public as-is, because people don't want e.g. push notifications while they're invisible, so we need to represent more states on the backend.